src/Controller/Controller.php

```diff
@@ -607,7 +607,12 @@ public function invokeAction()
         /* @var callable $callable */
         $callable = [$this, $request->getParam('action')];
 
-        return $callable(...array_values($request->getParam('pass')));
+        $result = $callable(...array_values($request->getParam('pass')));
+        if ($result instanceof Response) {
```

Comment:
Ideally this should check for `ResponseInterface` but I have kept it as `Response` since that's what `ActionDispatcher` also uses to check return value of various controller methods and event listeners.

We can try and change those to `ResponseInterface` checks in 4.0.